ldraw_to_scad.py

```diff
@@ -1,266 +1,303 @@
+                                   'FRACTION', 'VFRACTION', 'MATERIAL']:
+                            data[opt] = params[pos+4]
+                            skip = True
+                        elif opt in ['METAL', 'RUBBER', 'PEARLESCENT',
+                                     'CHROME']:
+                            data[opt] = True
+                        else:
+                            print(f'Unknown !COLOUR option {opt}!')
+                    alpha = int(data["ALPHA"]) if "ALPHA" in data else 255
+                    colors.append(
+                        f'(id=={data["CODE"]}) ? ["{data["VALUE"]}{alpha:02X}'
+                        f'","{data["EDGE"]}"] : (')
+            colors.append('"UNKNOWN"'+')'*len(colors)+';')
+            coltxt += '\n'.join(colors) + '\n'
+        return coltxt
 
     def index_library(self):
-        """Create an index of all LDRAW library items.
-        The index is a dictionary:
-        {"part_prefix\\part_name": "real_file_name.dat"},
-        eg:
-        {"1.dat": }
-        """
-        self.index = {}
-        paths = ['.']
-        for item in os.listdir('.'):
-            if item.endswith('.dat'):
-                self.index[item] = item
-
-        library_root = os.path.join('lib', 'ldraw')
-        for sub_path in ['parts', 'p']:
-            whole_path = os.path.join(library_root, sub_path) 
+        """ Index the whole library. """
+        index = {}
+        for sub_path in ['models', 'parts', 'p']:
+            whole_path = os.path.join(self.library_root, sub_path)
             for item in os.listdir(whole_path):
                 if item.endswith('.dat'):
-                    self.index[item] = os.path.join(whole_path, item)
+                    index[item] = [sub_path, os.path.splitext(item)[0]]
         special_subs = {
-            's': os.path.join(library_root, 'parts', 's'),
-            '48' : os.path.join(library_root, 'p', '48'),
-            '8' : os.path.join(library_root, 'p', '8')
+            's': os.path.join('parts', 's'),
+            '48': os.path.join('p', '48'),
+            '8': os.path.join('p', '8')
         }
         for prefix, s_path in special_subs.items():
-            for item in os.listdir(s_path):
+            for item in os.listdir(os.path.join(self.library_root, s_path)):
                 if item.endswith('.dat'):
-                    self.index[prefix + '\\'+item] = os.path.join(s_path, item)
+                    index[prefix + '\\'+item] = [s_path,
+                                                 os.path.splitext(item)[0]]
+        return index
 
     def find_part(self, part_name):
-        if not self.index:
-            self.index_library()
+        """ Find a part in the library. """
         try:
             filename = self.index[part_name]
         except KeyError:
             filename = self.index[part_name.lower()]
         return filename
 
+    def implementfunction(self, function):
```

Comment:
Can we snake_case these instead of running them together? readability counts.

Reply:
Coming from the C/C++ culture I tend to do things like this. But I hope I caught all cases now and changed as you suggested to properly adapt to the Python culture.